### PR TITLE
Use a more typesafe approach for managing indexed data

### PR DESCRIPTION
**Pull request by tifecool**
My attempt to resolve issue: #2403

Currently incomplete, having an error at polkadot-erasure-coding preventing me from testing, was hoping to get some advice and confirmation that I'm going about this the right way!

### CODE REVIEW

**Comment by tifecool** [comment]
@eskimor

File: primitives/src/v2/mod.rs
Comment (eskimor) [inline primitives/src/v2/mod.rs:151]: debug_assert would be good.

File: primitives/src/v2/mod.rs
Comment (eskimor) [inline primitives/src/v2/mod.rs:1627]: We are consuming the vector, there should be no need for clone.

File: runtime/parachains/src/inclusion/tests.rs
Comment (eskimor) [inline runtime/parachains/src/inclusion/tests.rs:1558]: Why not 2_usize? And why not have a `From` instance for i32?

**Review by eskimor — COMMENTED** [review]
Looks good at first pass.

**Comment by tifecool** [comment]
Thanks so much for the review! @eskimor 
Will keep at it and implement the suggested changes ❤️

File: runtime/parachains/src/inclusion/tests.rs
Comment (tifecool) [reply runtime/parachains/src/inclusion/tests.rs:1558]: Not sure but I think implementing a From instance for i32 would still lead to the same issue. The compiler was having issues deciding which number type to use since u32 and usize were both available.

How do I implement 2_usize (to usize?) (sorry about the beginner question)

File: primitives/src/v2/mod.rs
Comment (tifecool) [reply primitives/src/v2/mod.rs:151]: How would I implement this?

**Comment by ordian** [comment]
I'd vote for creating our own type-wrapper for `Vec` instead of using a dependency. We don't seem to need most of its methods other than type-safe indexing, which can be implemented easily as pointed out in https://github.com/paritytech/polkadot/issues/2403#issuecomment-777469483.

As a side note, please revert unrelated changes.

**Comment by tifecool** [comment]
@eskimor I fixed all the errors. Please help me check where the issues could be stemming from

**Comment by tifecool** [comment]
> I'd vote for creating our own type-wrapper for `Vec` instead of using a dependency. We don't seem to need most of its methods other than type-safe indexing, which can be implemented easily as pointed out in [#2403 (comment)](https://github.com/paritytech/polkadot/issues/2403#issuecomment-777469483).

Might need a bit of assistance/step through with this, but I think I have an idea of how to implement the change. Also, could I get help with why the checks are failing?

>As a side note, please revert unrelated changes.

Which changes?

**Comment by tifecool** [comment]
Not the best at git. Apologize for the numerous pushes

File: primitives/src/v2/mod.rs
Comment (drahnr) [inline primitives/src/v2/mod.rs:150]: Why do we need an impl for `From<usize>` if it's not guaranteed to be ok? Wouldn't a `From<u32>` not make more sense and leave it to the user to decide what to do on errors instead of generating a garbage `ValidatorIndex`? `TryFrom<usize>` and `From<u32>` seem more reasonable to me.

File: node/core/approval-voting/src/criteria.rs
Comment (drahnr) [inline node/core/approval-voting/src/criteria.rs:635]: we should consider adding a custom macro that resembles the ergonomy of `vec![..]`

File: primitives/src/v2/mod.rs
Comment (drahnr) [inline primitives/src/v2/mod.rs:1630]:
```suggestion
		for value in vec {
```

File: primitives/src/v2/mod.rs
Comment (drahnr) [inline primitives/src/v2/mod.rs:1633]:
```suggestion
                TypeVec::<K, V>::from_iterator(vec.into_iter())
```

File: primitives/src/v2/mod.rs
Comment (drahnr) [inline primitives/src/v2/mod.rs:1627]:
```suggestion
impl<K: From<usize>, V: Clone, I: IntoIterator<Item=V>> From<I> for TypeVec<K, V> {
```

File: primitives/src/v2/mod.rs
Comment (drahnr) [inline primitives/src/v2/mod.rs:1628]:
```suggestion
	fn from(vec: I) -> Self {
```

File: primitives/src/v2/mod.rs
Comment (drahnr) [inline primitives/src/v2/mod.rs:1671]: nit: Must be fixed before merge

**Review by drahnr — CHANGES_REQUESTED** [review]
I am not sure the chosen crate is the right choice. Having a fallback of `u32::MAX` is not necessarily a good thing when values are out of bounds when the only reason to have those bounds is the chosen crate to require `From<usize>` to be implemented.

File: primitives/src/v2/mod.rs
Comment (ordian) [inline primitives/src/v2/mod.rs:1631]: this is a very inefficient impl considering that `TiVec` impls `From<Vec<T>>`

File: primitives/src/v2/mod.rs
Comment (ordian) [inline primitives/src/v2/mod.rs:1640]: incorrect, should delegate to `Vec` impl

File: primitives/src/v2/mod.rs
Comment (ordian) [inline primitives/src/v2/mod.rs:1643]: incorrect, shouldn't be implemented

File: runtime/kusama/src/lib.rs
Comment (ordian) [inline runtime/kusama/src/lib.rs:450]: unrelated change (and probably doesn't pass rustfmt)

File: primitives/src/v2/mod.rs
Comment (ordian) [inline primitives/src/v2/mod.rs:1426]: please revert unrelated formatting changes, we're using a specific nightly version for fmt

File: erasure-coding/src/lib.rs
Comment (ordian) [inline erasure-coding/src/lib.rs:261]: unrelated change

**Comment by tifecool** [comment]
I'm getting an error anytime I try to run cargo test, how do I fix this:

```
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-Wl,--version-script=/tmp/rustc0dYdYt/list"
...
  = note: collect2: fatal error: ld terminated with signal 9 [Killed]
          compilation terminated.


error: could not compile `polkadot-cli` due to previous error

```

File: erasure-coding/src/lib.rs
Comment (tifecool) [reply erasure-coding/src/lib.rs:261]: I was getting this error:

```
error[E0282]: type annotations needed
   --> erasure-coding/src/lib.rs:261:28
    |
261 |         for (i, chunk) in chunks.as_ref().iter().enumerate() {
    |                           -------^^^^^^--
    |                           |      |
    |                           |      cannot infer type for type parameter `T` declared on the trait `AsRef`
    |                           this method call resolves to `&T`
    |
    = note: type must be known at this point
```

And on debugging saw the as_ref() made no difference to the object.

File: primitives/src/v2/mod.rs
Comment (tifecool) [reply primitives/src/v2/mod.rs:150]: Ok.

Comment (tifecool) [reply primitives/src/v2/mod.rs:150]: While working on this, the error:
```
the trait bound `K: From<usize>` is not satisfied
    --> primitives/src/v2/mod.rs:1616:17
     |
1616 | #[derive(Clone, RuntimeDebug)]
     |                 ^^^^^^^^^^^^ the trait `From<usize>` is not implemented for `K`
```
came up. This is why I likely implemented the `From<usize>` trait.

File: primitives/src/v2/mod.rs
Comment (tifecool) [reply primitives/src/v2/mod.rs:1671]: I'm not quite sure how to implement this trait.

File: primitives/src/v2/mod.rs
Comment (tifecool) [reply primitives/src/v2/mod.rs:1640]: Attempting this is causing issues.

Comment (tifecool) [reply primitives/src/v2/mod.rs:1640]: `#[cfg(feature = "std")]`
has to be implemented in numerous places

Comment (tifecool) [reply primitives/src/v2/mod.rs:1640]: Attempted implementation:
```
impl<K: From<usize>, V: Clone> MallocSizeOf for TypeVec<K, V> {
	fn size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
		self.to_vec().size_of(ops)
	}
}
```

**Comment by tifecool** [comment]
Going to reattempt this from scratch

**Pull request closed without merging**
